ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,336 @@
+      }
+    }
+  }
+
+  /**
+   * @return the current state of compaction.
+   */
+  State getCompactionState() {
+    return currentIdx >= cycleLogs.size() ? State.DONE : getCurrentCycleLog().getState();
+  }
+
+  /**
+   * @return the {@link CompactionDetails} for the compaction cycle in progress.
+   */
+  CompactionDetails getCompactionDetails() {
+    return getCurrentCycleLog().compactionDetails;
+  }
+
+  /**
+   * @return the {@link StoreFindToken} until which data has been copied and flushed.
+   */
+  StoreFindToken getSafeToken() {
+    return getCurrentCycleLog().safeToken;
+  }
+
+  /**
+   * Sets the {@link StoreFindToken} until which data is copied and flushed.
+   * @param safeToken the {@link StoreFindToken} until which data is copied and flushed.
+   */
+  void setSafeToken(StoreFindToken safeToken) {
+    CycleLog cycleLog = getCurrentCycleLog();
+    if (!cycleLog.getState().equals(State.COPY)) {
+      throw new IllegalStateException("Cannot set a safe token - not in COPY phase");
+    }
+    cycleLog.safeToken = safeToken;
+    flush();
+  }
+
+  /**
+   * Marks the start of the copy phase.
+   */
+  void markCopyStart() {
+    getCurrentCycleLog().copyStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the switch phase.
+   */
+  void markSwitchStart() {
+    getCurrentCycleLog().switchStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the start of the cleanup phase.
+   */
+  void markCleanupStart() {
+    getCurrentCycleLog().cleanupStartTime = time.milliseconds();
+    flush();
+  }
+
+  /**
+   * Marks the current compaction cycle as complete.
+   */
+  void markCycleComplete() {
+    getCurrentCycleLog().cycleEndTime = time.milliseconds();
+    currentIdx++;
+    flush();
+  }
+
+  /**
+   * Closes the compaction log. If compaction is complete, renames the log to keep a permanent record.
+   */
+  @Override
+  public void close() {
+    if (file.exists() && getCompactionState().equals(State.DONE)) {
+      File savedLog = new File(file.getAbsolutePath() + "_" + startTime);
+      if (!file.renameTo(savedLog)) {
+        throw new IllegalStateException("Compaction log could not be renamed after completion of compaction");
+      }
+    }
+  }
+
+  /**
+   * @return the {@link CycleLog} for the current compaction cycle.
+   */
+  private CycleLog getCurrentCycleLog() {
+    return cycleLogs.get(currentIdx);
+  }
+
+  /**
+   * Flushes all changes to the file backing this compaction log.
+   */
+  private void flush() {
+    /*
+        Description of serialized format
+
+        version
+        startTime
+        index of current cycle's log
+        size of cycle log list
+        cycleLog1 (see CycleLog#toBytes())
+        cycleLog2
+        ...
+        crc
+       */
+    File tempFile = new File(file.getAbsolutePath() + ".tmp");
+    try (FileOutputStream fileOutputStream = new FileOutputStream(tempFile)) {
```

Comment:
Could you clarify why you think things will fail if the ".tmp" file exists? `FileOutputStream` does not throw an error if the file already exists (just verified it also).